controllers/pulp/ingress.go

```diff
@@ -0,0 +1,212 @@
+			TargetPort:  "content-24816",
+			ServiceName: pulp.Name + "-content-svc",
+		},
+		{
+			Name:        pulp.Name + "-api-v3",
+			Path:        getPulpSetting(pulp, "api_root") + "api/v3/",
+			TargetPort:  "api-24817",
+			ServiceName: pulp.Name + "-api-svc",
+		},
+		{
+			Name:        pulp.Name + "-auth",
+			Path:        "/auth/login/",
+			TargetPort:  "api-24817",
+			ServiceName: pulp.Name + "-api-svc",
+		},
+		{
+			Name:        pulp.Name,
+			Path:        "/",
+			TargetPort:  "api-24817",
+			ServiceName: pulp.Name + "-api-svc",
+		},
+	}
+	pulpPlugins = append(defaultPlugins, pulpPlugins...)
+
+	// get ingress
+	pulpIngress := &netv1.Ingress{}
+	err = r.Get(ctx, types.NamespacedName{Name: pulp.Name, Namespace: pulp.Namespace}, pulpIngress)
+
+	// Create the ingress in case it is not found
+	if err != nil && errors.IsNotFound(err) {
+		ingressObj := pulpIngressObject(pulp, pulpPlugins)
+		ctrl.SetControllerReference(pulp, ingressObj, r.Scheme)
+		log.Info("Creating a new ingress", "Ingress.Namespace", ingressObj.Namespace, "Ingress.Name", ingressObj.Name)
+		r.updateStatus(ctx, pulp, metav1.ConditionFalse, pulp.Spec.DeploymentType+"-Ingress-Ready", "CreatingIngress", "Creating "+pulp.Name+"-ingress")
+		err = r.Create(ctx, ingressObj)
+		if err != nil {
+			log.Error(err, "Failed to create new ingress", "Ingress.Namespace", ingressObj.Namespace, "Ingress.Name", ingressObj.Name)
+			r.updateStatus(ctx, pulp, metav1.ConditionFalse, pulp.Spec.DeploymentType+"-Ingress-Ready", "ErrorCreatingIngress", "Failed to create "+pulp.Name+"-ingress: "+err.Error())
+			r.recorder.Event(pulp, corev1.EventTypeWarning, "Failed", "Failed to create new ingress")
+			return ctrl.Result{}, err
+		}
+	} else if err != nil {
+		log.Error(err, "Failed to get ingress")
+		return ctrl.Result{}, err
+	}
+
+	// we should only update the status when Ingress-Ready==false
+	if v1.IsStatusConditionFalse(pulp.Status.Conditions, cases.Title(language.English, cases.Compact).String(pulp.Spec.DeploymentType)+"-Ingress-Ready") {
+		r.updateStatus(ctx, pulp, metav1.ConditionTrue, pulp.Spec.DeploymentType+"-Ingress-Ready", "IngressTasksFinished", "All Ingress tasks ran successfully")
+		r.recorder.Event(pulp, corev1.EventTypeNormal, "IngressReady", "All Ingress tasks ran successfully")
+	}
+	return ctrl.Result{}, nil
+}
+
+// pulp-ingress
+func pulpIngressObject(m *repomanagerv1alpha1.Pulp, plugins []IngressPlugin) *netv1.Ingress {
+	annotation := map[string]string{
+		"haproxy.router.openshift.io/timeout":               m.Spec.HAProxyTimeout,
+		"nginx.ingress.kubernetes.io/proxy-body-size":       m.Spec.NginxProxyBodySize,
+		"nginx.org/client-max-body-size":                    m.Spec.NginxMaxBodySize,
+		"nginx.ingress.kubernetes.io/proxy-read-timeout":    m.Spec.NginxProxyReadTimeout,
+		"nginx.ingress.kubernetes.io/proxy-connect-timeout": m.Spec.NginxProxyConnectTimeout,
+		"nginx.ingress.kubernetes.io/proxy-send-timeout":    m.Spec.NginxProxySendTimeout,
```

Comment:
Hum ... what if we let these annotations as a "generic" variable/field?
My concern is if someone has a Big-IP, or an Istio, or a Citrix instead of Nginx as an ingress controller  (https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/) we would need to cover all of these possibilities.

Reply:
I didn't know about this other possibilities, I'm wondering if the redirect annotation would work for them

Reply:
I just learned the rewrite doesn't work on other ingress controllers:
https://stackoverflow.com/a/62086838/5253051

Reply:
I added a new var: `ingress_class` and its default is `nginx`, so if it is a nginx ingress we use the annotations, if not the ingress will use the pulp-web service.

It seems the fallback for ingress that doesn't support redirect is using nginx: https://github.com/kubernetes/ingress-gce/issues/109#issuecomment-1120472204
as we already have pulp-web for this job, that's why I'm using the pulp-web service when the ingress isn't a nginx one.

I plan to document it in a follow-up PR

Reply:
actually I added the doc on reverse proxy section

Reply:
> It seems the fallback for ingress that doesn't support redirect is using nginx: https://github.com/kubernetes/ingress-gce/issues/109#issuecomment-1120472204
as we already have pulp-web for this job, that's why I'm using the pulp-web service when the ingress isn't a nginx one.

Hum... it seems like a deficit in cloud providers (aws/gcp) load balancer, right?
I believe that some vendors (like F5 Big IP, Citrix, HAProxy) do have the rewrite implemented in their ingress controllers.

Reply:
the problem would be having a switch case for all the different annotations